// secretjs/build/cosmwasmclient.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/camelcase */
const crypto_1 = require("@iov/crypto");
const encoding_1 = require("@iov/encoding");
const utils_1 = require("@iov/utils");
const readonly_date_1 = require("readonly-date");
const cosmwasmclient_1 = require("./cosmwasmclient");
const encoding_2 = require("./encoding");
const logs_1 = require("./logs");
const pen_1 = require("./pen");
const signingcosmwasmclient_1 = require("./signingcosmwasmclient");
const cosmoshub_json_1 = __importDefault(require("./testdata/cosmoshub.json"));
const testutils_spec_1 = require("./testutils.spec");
const { fromHex, fromUtf8, toAscii, toBase64 } = encoding_1.Encoding;
const guest = {
    address: "cosmos17d0jcz59jf68g52vq38tuuncmwwjk42u6mcxej",
};
describe("CosmWasmClient", () => {
    describe("makeReadOnly", () => {
        it("can be constructed", () => {
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            expect(client).toBeTruthy();
        });
    });
    describe("getChainId", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            expect(await client.getChainId()).toEqual(testutils_spec_1.wasmd.chainId);
        });
        it("caches chain ID", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const openedClient = client;
            const getCodeSpy = spyOn(openedClient.restClient, "nodeInfo").and.callThrough();
            expect(await client.getChainId()).toEqual(testutils_spec_1.wasmd.chainId); // from network
            expect(await client.getChainId()).toEqual(testutils_spec_1.wasmd.chainId); // from cache
            expect(getCodeSpy).toHaveBeenCalledTimes(1);
        });
    });
    describe("getHeight", () => {
        it("gets height via last block", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const openedClient = client;
            const blockLatestSpy = spyOn(openedClient.restClient, "blocksLatest").and.callThrough();
            const height1 = await client.getHeight();
            expect(height1).toBeGreaterThan(0);
            await utils_1.sleep(1000);
            const height2 = await client.getHeight();
            expect(height2).toEqual(height1 + 1);
            expect(blockLatestSpy).toHaveBeenCalledTimes(2);
        });
        it("gets height via authAccount once an address is known", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const openedClient = client;
            const blockLatestSpy = spyOn(openedClient.restClient, "blocksLatest").and.callThrough();
            const authAccountsSpy = spyOn(openedClient.restClient, "authAccounts").and.callThrough();
            const height1 = await client.getHeight();
            expect(height1).toBeGreaterThan(0);
            await client.getCodes(); // warm up the client
            const height2 = await client.getHeight();
            expect(height2).toBeGreaterThan(0);
            await utils_1.sleep(1000);
            const height3 = await client.getHeight();
            expect(height3).toEqual(height2 + 1);
            expect(blockLatestSpy).toHaveBeenCalledTimes(1);
            expect(authAccountsSpy).toHaveBeenCalledTimes(2);
        });
    });
    describe("getNonce", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            expect(await client.getNonce(testutils_spec_1.unused.address)).toEqual({
                accountNumber: testutils_spec_1.unused.accountNumber,
                sequence: testutils_spec_1.unused.sequence,
            });
        });
        it("throws for missing accounts", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const missing = testutils_spec_1.makeRandomAddress();
            await client.getNonce(missing).then(() => fail("this must not succeed"), (error) => expect(error).toMatch(/account does not exist on chain/i));
        });
    });
    describe("getAccount", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            expect(await client.getAccount(testutils_spec_1.unused.address)).toEqual({
                address: testutils_spec_1.unused.address,
                accountNumber: testutils_spec_1.unused.accountNumber,
                sequence: testutils_spec_1.unused.sequence,
                pubkey: undefined,
                balance: [
                    { denom: "ucosm", amount: "1000000000" },
                    { denom: "ustake", amount: "1000000000" },
                ],
            });
        });
        it("returns undefined for missing accounts", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const missing = testutils_spec_1.makeRandomAddress();
            expect(await client.getAccount(missing)).toBeUndefined();
        });
    });
    describe("getBlock", () => {
        it("works for latest block", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const response = await client.getBlock();
            // id
            expect(response.id).toMatch(testutils_spec_1.tendermintIdMatcher);
            // header
            expect(response.header.height).toBeGreaterThanOrEqual(1);
            expect(response.header.chainId).toEqual(await client.getChainId());
            expect(new readonly_date_1.ReadonlyDate(response.header.time).getTime()).toBeLessThan(readonly_date_1.ReadonlyDate.now());
            expect(new readonly_date_1.ReadonlyDate(response.header.time).getTime()).toBeGreaterThanOrEqual(readonly_date_1.ReadonlyDate.now() - 5000);
            // txs
            expect(Array.isArray(response.txs)).toEqual(true);
        });
        it("works for block by height", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const height = (await client.getBlock()).header.height;
            const response = await client.getBlock(height - 1);
            // id
            expect(response.id).toMatch(testutils_spec_1.tendermintIdMatcher);
            // header
            expect(response.header.height).toEqual(height - 1);
            expect(response.header.chainId).toEqual(await client.getChainId());
            expect(new readonly_date_1.ReadonlyDate(response.header.time).getTime()).toBeLessThan(readonly_date_1.ReadonlyDate.now());
            expect(new readonly_date_1.ReadonlyDate(response.header.time).getTime()).toBeGreaterThanOrEqual(readonly_date_1.ReadonlyDate.now() - 5000);
            // txs
            expect(Array.isArray(response.txs)).toEqual(true);
        });
    });
    describe("getIdentifier", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            expect(await client.getIdentifier(cosmoshub_json_1.default.tx)).toEqual(cosmoshub_json_1.default.id);
        });
    });
    describe("postTx", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const pen = await pen_1.Secp256k1Pen.fromMnemonic(testutils_spec_1.faucet.mnemonic);
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const memo = "My first contract on chain";
            const sendMsg = {
                type: "cosmos-sdk/MsgSend",
                value: {
                    from_address: testutils_spec_1.faucet.address,
                    to_address: testutils_spec_1.makeRandomAddress(),
                    amount: [
                        {
                            denom: "ucosm",
                            amount: "1234567",
                        },
                    ],
                },
            };
            const fee = {
                amount: [
                    {
                        amount: "5000",
                        denom: "ucosm",
                    },
                ],
                gas: "890000",
            };
            const chainId = await client.getChainId();
            const { accountNumber, sequence } = await client.getNonce(testutils_spec_1.faucet.address);
            const signBytes = encoding_2.makeSignBytes([sendMsg], fee, chainId, memo, accountNumber, sequence);
            const signature = await pen.sign(signBytes);
            const signedTx = {
                msg: [sendMsg],
                fee: fee,
                memo: memo,
                signatures: [signature],
            };
            const { logs, transactionHash } = await client.postTx(signedTx);
            const amountAttr = logs_1.findAttribute(logs, "transfer", "amount");
            expect(amountAttr.value).toEqual("1234567ucosm");
            expect(transactionHash).toMatch(/^[0-9A-F]{64}$/);
        });
    });
    describe("getCodes", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const result = await client.getCodes();
            expect(result.length).toBeGreaterThanOrEqual(1);
            const [first] = result;
            expect(first).toEqual({
                id: testutils_spec_1.deployedErc20.codeId,
                source: testutils_spec_1.deployedErc20.source,
                builder: testutils_spec_1.deployedErc20.builder,
                checksum: testutils_spec_1.deployedErc20.checksum,
                creator: testutils_spec_1.faucet.address,
            });
        });
    });
    describe("getCodeDetails", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const result = await client.getCodeDetails(1);
            const expectedInfo = {
                id: testutils_spec_1.deployedErc20.codeId,
                source: testutils_spec_1.deployedErc20.source,
                builder: testutils_spec_1.deployedErc20.builder,
                checksum: testutils_spec_1.deployedErc20.checksum,
                creator: testutils_spec_1.faucet.address,
            };
            // check info
            expect(result).toEqual(jasmine.objectContaining(expectedInfo));
            // check data
            expect(new crypto_1.Sha256(result.data).digest()).toEqual(fromHex(expectedInfo.checksum));
        });
        it("caches downloads", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const openedClient = client;
            const getCodeSpy = spyOn(openedClient.restClient, "getCode").and.callThrough();
            const result1 = await client.getCodeDetails(testutils_spec_1.deployedErc20.codeId); // from network
            const result2 = await client.getCodeDetails(testutils_spec_1.deployedErc20.codeId); // from cache
            expect(result2).toEqual(result1);
            expect(getCodeSpy).toHaveBeenCalledTimes(1);
        });
    });
    describe("getContracts", () => {
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const result = await client.getContracts(1);
            expect(result.length).toBeGreaterThanOrEqual(3);
            const [hash, isa, jade] = result;
            expect(hash).toEqual({
                address: "cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5",
                codeId: 1,
                creator: testutils_spec_1.faucet.address,
                label: "HASH",
            });
            expect(isa).toEqual({
                address: "cosmos1hqrdl6wstt8qzshwc6mrumpjk9338k0lr4dqxd",
                codeId: 1,
                creator: testutils_spec_1.faucet.address,
                label: "ISA",
            });
            expect(jade).toEqual({
                address: "cosmos18r5szma8hm93pvx6lwpjwyxruw27e0k5uw835c",
                codeId: 1,
                creator: testutils_spec_1.faucet.address,
                label: "JADE",
            });
        });
    });
    describe("getContract", () => {
        it("works for HASH instance", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const hash = await client.getContract("cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5");
            expect(hash).toEqual({
                address: "cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5",
                codeId: 1,
                creator: testutils_spec_1.faucet.address,
                label: "HASH",
                initMsg: {
                    decimals: 5,
                    name: "Hash token",
                    symbol: "HASH",
                    initial_balances: jasmine.arrayContaining([
                        {
                            address: testutils_spec_1.faucet.address,
                            amount: "11",
                        },
                        {
                            address: testutils_spec_1.unused.address,
                            amount: "12812345",
                        },
                        {
                            address: guest.address,
                            amount: "22004000000",
                        },
                    ]),
                },
            });
        });
    });
    describe("queryContractRaw", () => {
        const configKey = toAscii("config");
        const otherKey = toAscii("this_does_not_exist");
        let contract;
        beforeAll(async () => {
            if (testutils_spec_1.wasmdEnabled()) {
                testutils_spec_1.pendingWithoutWasmd();
                const pen = await pen_1.Secp256k1Pen.fromMnemonic(testutils_spec_1.faucet.mnemonic);
                const client = new signingcosmwasmclient_1.SigningCosmWasmClient(testutils_spec_1.wasmd.endpoint, testutils_spec_1.faucet.address, (signBytes) => pen.sign(signBytes));
                const { codeId } = await client.upload(testutils_spec_1.getHackatom());
                const initMsg = { verifier: testutils_spec_1.makeRandomAddress(), beneficiary: testutils_spec_1.makeRandomAddress() };
                const { contractAddress } = await client.instantiate(codeId, initMsg, "random hackatom");
                contract = { initMsg: initMsg, address: contractAddress };
            }
        });
    });
    describe("queryContractSmart", () => {
        let contract;
        beforeAll(async () => {
            if (testutils_spec_1.wasmdEnabled()) {
                testutils_spec_1.pendingWithoutWasmd();
                const pen = await pen_1.Secp256k1Pen.fromMnemonic(testutils_spec_1.faucet.mnemonic);
                const client = new signingcosmwasmclient_1.SigningCosmWasmClient(testutils_spec_1.wasmd.endpoint, testutils_spec_1.faucet.address, (signBytes) => pen.sign(signBytes));
                const { codeId } = await client.upload(testutils_spec_1.getHackatom());
                const initMsg = { verifier: testutils_spec_1.makeRandomAddress(), beneficiary: testutils_spec_1.makeRandomAddress() };
                const { contractAddress } = await client.instantiate(codeId, initMsg, "a different hackatom");
                contract = { initMsg: initMsg, address: contractAddress };
            }
        });
        it("works", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            utils_1.assert(contract);
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            const resultDocument = await client.queryContractSmart(contract.address, { verifier: {} });
            expect(resultDocument).toEqual({ verifier: contract.initMsg.verifier });
        });
        it("errors for malformed query message", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            utils_1.assert(contract);
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            await client.queryContractSmart(contract.address, { broken: {} }).then(() => fail("must not succeed"), (error) => expect(error).toMatch(/query contract failed: parsing hackatom::contract::QueryMsg/i));
        });
        it("errors for non-existent contract", async () => {
            testutils_spec_1.pendingWithoutWasmd();
            const nonExistentAddress = testutils_spec_1.makeRandomAddress();
            const client = new cosmwasmclient_1.CosmWasmClient(testutils_spec_1.wasmd.endpoint);
            await client.queryContractSmart(nonExistentAddress, { verifier: {} }).then(() => fail("must not succeed"), (error) => expect(error).toMatch(`No contract found at address "${nonExistentAddress}"`));
        });
    });
});
