// @iov/encoding/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Re-exports for IOV Core 2.x compatibility
var math_1 = require("@cosmjs/math");
exports.Int53 = math_1.Int53;
exports.Uint32 = math_1.Uint32;
exports.Uint53 = math_1.Uint53;
exports.Uint64 = math_1.Uint64;
exports.Decimal = math_1.Decimal;
var encoding_1 = require("@cosmjs/encoding");
exports.fromHex = encoding_1.fromHex;
exports.toHex = encoding_1.toHex;
exports.fromAscii = encoding_1.fromAscii;
exports.toAscii = encoding_1.toAscii;
exports.fromBase64 = encoding_1.fromBase64;
exports.toBase64 = encoding_1.toBase64;
exports.fromRfc3339 = encoding_1.fromRfc3339;
exports.toRfc3339 = encoding_1.toRfc3339;
exports.fromUtf8 = encoding_1.fromUtf8;
exports.toUtf8 = encoding_1.toUtf8;
exports.Bech32 = encoding_1.Bech32;
var utils_1 = require("@cosmjs/utils");
exports.isNonNullObject = utils_1.isNonNullObject;
exports.isUint8Array = utils_1.isUint8Array;
// Symbols not (yet) existent in CosmJS
var json_1 = require("./json");
exports.isJsonCompatibleValue = json_1.isJsonCompatibleValue;
exports.isJsonCompatibleDictionary = json_1.isJsonCompatibleDictionary;
exports.isJsonCompatibleArray = json_1.isJsonCompatibleArray;
var transactionencoder_1 = require("./transactionencoder");
exports.TransactionEncoder = transactionencoder_1.TransactionEncoder;
// Deprecated symbols to be removed in the next mayor release
var encoding_2 = require("./encoding");
exports.Encoding = encoding_2.Encoding;
